Delegate properly to custom Resolver instances

bquorning opened the pull request:
Let `EmptyTemplateResolver` creating instances of either `EmptyTemplateResolver::ResolverDecorator` or `EmptyTemplateResolver::FileSystemResolver`, depending on whether the given path is an `ActionView::Resolver` instance or not.

The new `EmptyTemplateResolver::ResolverDecorator` class simply delegates all method calls to its given resolver, except for `#find_templates` which it overrides to return templates that render with `EmptyTemplateHandler`.

Please note that the merge target is PR #1557, not master.

bquorning commented:
/cc @JonRowe @samphippen @pixeltrix

samphippen commented:
LGTM, but I'd like another review.

JonRowe commented on lib/rspec/rails/view_rendering.rb:
```diff
@@ -39,18 +39,50 @@ def render_views?
-      #
-      # @private
-      class EmptyTemplateResolver < ::ActionView::FileSystemResolver
+      class EmptyTemplateResolverFactory
```
When we build thinks like this we normally make them "instance less" to avoid unnecessary allocations, please refactor to a static instance, or better yet a module method. Can I also suggest a better name is desired? Something like `DecoratePath` or `BuildEmptyTemplateResolver`

bquorning replied:
Agree on both the allocation and naming concerns. How does `EmptyTemplateResolver.build` sound?

bquorning replied:
Actually, we could move `EmptyTemplateFileSystemResolver` and `EmptyTemplateResolverDecorator` into this `EmptyTemplateResolver` module (class), since they’re already supposed to be private and are only used by this factory class.

bquorning commented on lib/rspec/rails/view_rendering.rb:
```diff
@@ -39,18 +39,50 @@ def render_views?
         end
+      end
 
+      class EmptyTemplateResolverDecorator
```
```
# Delegates find_templates to the submitted resolver and then returns templates
# with modified source
#
# @private
```

bquorning commented on lib/rspec/rails/view_rendering.rb:
```diff
@@ -39,18 +39,50 @@ def render_views?
+        end
+      end
+
+      # Delegates find_all to the submitted path set and then returns templates
```
Should be `find_templates`.

JonRowe commented on lib/rspec/rails/view_rendering.rb:
```diff
@@ -39,18 +39,50 @@ def render_views?
+
+        def find_templates(*args)
+          templates = @resolver.find_templates(*args)
+          templates.map do |template|
```
This map could probably now be delegated to outside as it does the same as below.

bquorning replied:
Yeah, but I don’t really know where to move it. `EmptyTemplateHandler`?

JonRowe commented on lib/rspec/rails/view_rendering.rb:
```diff
@@ -88,13 +120,13 @@ def append_view_path(new_path)
 
         def _path_decorator(*paths)
-          paths.map { |path| EmptyTemplateResolver.new(path) }
+          paths.map { |path| EmptyTemplateResolverFactory.new(path).resolver }
```
See above about allocations.

bquorning commented:
There, I pushed some changes:

 1. Changed the factory from an instance method to a class method: `EmptyTemplateResolver.build`
 2. Renamed and moved the `ResolverDecorator` and `FileSystemResolver` classes into `EmptyTemplateResolver::` namespace.
 3. This allowed for extracting the duplicated mapping into an `EmptyTemplateResolver.nullify_template_rendering` method.

samphippen commented:
@JonRowe LGTM

samphippen commented:
Why are there no travis results for this branch?

samphippen commented:
(did not mean to close)

bquorning commented:
> Why are there no travis results for this branch?

Perhaps because the merge target is not master branch?

JonRowe commented:
Can you provide an example of the code this is intended to fix? I'm still hesitant around this as I haven't seen a real world scenario that's broken by the existing code...

bquorning commented:
@JonRowe I created a gist at https://gist.github.com/bquorning/6bf534e4b145b1738bb2. It’s a bit contrived, since I didn’t want to create tables and fetch custom templates from the database.

ento commented:
re: example:
I have a group of namespaced controllers that share the same templates and a module of shared actions. I believe this PR will fix my anonymous controller test for the module, which uses [a custom resolver](http://blog.plataformatec.com.br/2011/04/default-views-in-rails-3-0-with-custom-resolvers/) to get to those templates.

bquorning commented:
By the way, should I rather make a PR against master, so that we can see some Travis runs?

bquorning commented:
Added 1 commit that allows Travis to run. Please don’t merge the PR with this commit included.

bquorning commented:
Build fails because of the new version of Rake that has [Removed to support Ruby 1.8.x](https://github.com/ruby/rake/blob/master/History.rdoc#1100--2016-03-09)

bquorning commented:
Superseded by #1580.
